tachyon/zk/air/constraint_system/constraint_system.h

```diff
@@ -0,0 +1,140 @@
+#ifndef TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+#define TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+
+#include <stddef.h>
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "third_party/eigen3/Eigen/Core"
+
+#include "tachyon/zk/air/constraint_system/variable.h"
+#include "tachyon/zk/air/expressions/evaluator.h"
+
+namespace tachyon::zk::air {
+
+template <typename F>
+class ConstraintSystem {
+ public:
+  ConstraintSystem() = default;
+  ConstraintSystem(size_t preprocessed_width, size_t main_width,
+                   size_t num_public_values)
+      : preprocessed_width_(preprocessed_width),
+        main_width_(main_width),
+        num_public_values_(num_public_values) {
+    preprocessed_.reserve(preprocessed_width * 2);
+    main_.reserve(main_width * 2);
+    public_values_.reserve(num_public_values);
+
+    for (size_t row : {0, 1}) {
+      for (Eigen::Index i = 0; i < preprocessed_width_; ++i) {
+        preprocessed_.push_back(
+            Variable::Preprocessed(row * preprocessed_width_ + i));
+      }
+    }
+    for (size_t row : {0, 1}) {
+      for (Eigen::Index i = 0; i < main_width_; ++i) {
+        main_.push_back(Variable::Main(row * main_width_ + i));
+      }
+    }
+    for (size_t i = 0; i < num_public_values; ++i) {
+      public_values_.push_back(Variable::Public(i));
+    }
+  }
+
+  const std::vector<Variable>& GetPublicValues() const {
+    return public_values_;
+  }
```

Comment:
In our convention, it seems more appropriate to name this function `public_values()`.